File: src/frontend/position.rs
#[derive(Debug, Clone)]
/// A coordinate implementation for simplified column and line.
pub struct Position {
    /// The column position
    pub column: i32,
    /// The initial column position
    pub initial_column: i32,
    /// The line position
    pub line: i32,
}

impl Position {
    pub fn new(column: i32, line: i32) -> Self {
        Position { column, initial_column: column, line }
    }

    pub fn next_column(&mut self) {
        self.column += 1;
    }

    pub fn next_line(&mut self) {
        self.line += 1;
        self.column = self.initial_column;
    }
}

File: src/frontend/lexer/Position.ts
export class Position {
    public column: number;
    public line: number;

    constructor(column: number, line: number) {
        this.column = column;
        this.line = line;
    }

    public nextColumn(): void {
        this.column += 1;
    }

    /** Also resets the column. */
    public nextLine(column: number = 1): void {
        this.line += 1;
        this.column = column;
    }

    public clone(): { column: number; line: number } {
        return { column: this.column, line: this.line };
    }
}
